### ISSUE TITLE
Default rules *way* too strict

### ISSUE THREAD

**Issue by devinus** [jonathanKingston/eslint-config-ember#2]
I really want to use this, but the default rules are *way* too strict. Simple `ember-cli` projects don't pass without significant changes.

I feel like if you're going to take `ember` as the namespace it should work to serve the majority of Ember projects and perhaps adopt a community standard of rules.

**Comment by green-arrow** [comment]
I agree with @devinus on this. Also, many new ember projects embrace things like function expressions, which are disallowed by default. It's easy enough to override, I think the config should at least work with a freshly generated ember-cli app. Or maybe the ember-cli blueprints need to change?

**Comment by jonathanKingston** [comment]
I actually agree... however I was waiting for a discussion like this. I was preparing to update the package but have not had time. The main fault is actually the sub package nightmare-mode which can be as nasty as it likes.

The aim of this package is to be defacto Ember so it passes on native Ember code and the latest examples etc whilst still being opinionated enough to fix some of the issues that people cause for themselves.

Function expressions should be fixed as it is definitely common Ember - despite the shorthand being cleaner (which I think is permitted):

```
{
  thing() {
  }
}
```

I agree on all the points. I think also that there should be more examples of how to fix the issues here.

Sorry for the delays etc. (I do take pull requests if rationale is put in place etc)

**Comment by jonathanKingston** [comment]
Hey @devinus and @green-arrow

So here is a quick update where I am at so far.

I have updated eslint-config-nightmare-mode which this project depends upon: https://github.com/jonathanKingston/eslint-config-nightmare-mode/commit/365b96b0346f6a0b18de21816c0e7ee7e2198ecd

This introduces a new testing format which is far simpler (which in itself is new and shouldn't really be depended on just yet) once I am convinced this is working I want to start using Ember blueprint outputs as example code to test here which should have no errors as you mention. This means we can get dependable tests that cope with changes from both ESLint and nightmare-mode whilst getting the strictest set possible without annoying too many people.

After I get this in place it would be great for everyone to review and ensure they are happy with it.

**Comment by devinus** [comment]
@jonathanKingston I think taking Ember's own style guide and maybe Dockyard's into consideration would be a good starting point as well.

https://github.com/emberjs/ember.js/blob/master/STYLEGUIDE.md
https://github.com/dockyard/styleguides/blob/master/engineering/ember.md

Personally, I've moved to using Airbnb's great ES6 style guide. It's the closest to what I've seen the vast majority of the ES6 ecosystem move to lately.

https://github.com/airbnb/javascript

**Comment by jonathanKingston** [comment]
@nickiaconis I think we need some contributions here to help shape this into a real commit. Given your work I would like to see your feedback :+1:

**Comment by BrianSipple** [comment]
I like the idea of extending from a “base” style guide and layering Ember-specific rules on top. As @devinus mentioned, Airbnb's styleguide is pretty fantastic, and I'm also intrigued by what `eslint-config-nightmare-mode` could offer us. 

Furthermore -- and this is just a bit of brainstorming -- I'm wondering if the rules that we need to extend in `eslint-config-ember` could serve as “trackers” of sorts -- possible beneficial changes that could be made to Ember-CLI itself in order to make the code generated by its blueprints more compatible with the community’s most popular/best-practice linting configurations. That would be a much larger discussion to probably bring some core team members in on -- but just a thought.

Anyway, for what it's worth, these are the extensions that I found necessary for getting my tests to pass ESLint after implementing an root-level config that first extended `eslint-config-ember`, and then extended a custom ruleset very similar to the AirBnB style guide. A few -- like `max-len` and `prefer-reflect` are more reflective of my own preferences, but most of these are direct responses to Ember-CLI's generated stubs:

```
  // overrides of our primary ESLint settings that are specifically targeted to
  // the Ember-CLI testing environment.
  "rules": {
    "arrow-spacing": 0,
    "comma-dangle": 0,
    "func-names": 0,
    "max-len": [1, 110, 2, {
      "ignoreComments": true,
      "ignoreUrls": true,
      "ignorePattern": "^moduleFor\(\?\:Component)\?\\(.*"
    }],
    "object-shorthand": 0,
    "newline-after-var": 0,
    "no-console": 0,
    "no-invalid-this": 0,
    "object-curly-spacing": [2, "always"],
    "prefer-arrow-callback": 0,
    "prefer-const": 0,
    "prefer-reflect": 0,
    "space-before-function-paren": [0, { "anonymous": "always", "named": "always" }]
}
```

Needless to say, I'd definitely be willing to help out with how we set things up going forward.

**Comment by jonathanKingston** [comment]
The initial rationale for `nightmare-mode` was to give you one of the strictest setups possible to change configs to only be a whitelist of bits that should be permitted.

The Airbnb's is fantastic yeah however without the documentation it is a little lost, I did some brainstorming about auto generating that based upon rules used but it never really got further than that.

> Furthermore -- and this is just a bit of brainstorming -- I'm wondering if the rules that we need to extend in eslint-config-ember could serve as “trackers” of sorts

That was the initial idea I had myself. Ideally using unit tests like [eslint-code-review](https://github.com/jonathanKingston/eslint-code-review) or the internal test suite to check the blueprint outputs. There was plans to write our own rules that made sense to Ember... of which there would be loads that I'm sure would make sense like teaching bad practices away from beginners. As @Turbo87 has been worked on bringing ESLint to ember-cli I think taking his work and pushing it in here would probably make sense too.

Yeah prefer-reflect comes out of nightmare mode however limits issues in applying to the wrong `this`.
Looks like some of these rules would be better in a separate test config file which could then be included like:
```
{
  "extends": "ember/test"
}
```

Thanks for writing this! I'm glad you are willing to be helping, as you can probably tell my work on Ember has slowed a lot however I am still very keen on linting.

I suggest more [tests like this](https://github.com/jonathanKingston/eslint-config-nightmare-mode/blob/master/test/index.js) would prevent future unexpected strict jumps when nightmare mode gets stricter.

**Comment by jonathanKingston** [comment]
Yup I saw them a while back, last time I looked at them they didn't seem very helpful. Happy to revisit if they are.

**Comment by jonathanKingston** [comment]
@Turbo87 can we move this over to ember-cli org and fix the issues there? As @BrianSipple said the separation this gives is slightly neater than just giving developers a default policy.
